components/codec/src/byte.rs

```diff
@@ -352,12 +378,172 @@ impl MemComparableCodecHelper for DescendingMemComparableCodecHelper {
+        unsafe {
+            self.advance_mut(len);
+        }
+        Ok(())
+    }
+
+    /// write all bytes in descending memory-comparable format.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::BufferTooSmall` if buffer remaining size < values.len().
+    fn write_bytes_desc(&mut self, bs: &[u8]) -> Result<()> {
+        let len = MemComparableByteCodec::encoded_len(bs.len());
+        let buf = unsafe { self.bytes_mut(len) };
+        if buf.len() < len {
+            return Err(Error::BufferTooSmall);
+        }
+        MemComparableByteCodec::encode_all_desc(bs, buf);
+        unsafe {
+            self.advance_mut(len);
+        }
+        Ok(())
+    }
+}
+
+impl<T: NumberEncoder> MemComparableByteEncoder for T {}
+
+pub trait MemComparableByteDecoder: BufferReader {
+    fn read_bytes(&mut self) -> Result<Vec<u8>> {
+        let mut buf = vec![0; self.bytes().len()];
+        let (read, written) = MemComparableByteCodec::try_decode_first(self.bytes(), &mut buf)?;
+        self.advance(read);
+        Ok(buf[..written].to_vec())
+    }
+}
+
+impl<T: BufferReader> MemComparableByteDecoder for T {}
+
+pub trait CompactByteEncoder {
+    /// Joins bytes with its length into a byte slice.
+    /// It is more efficient in both space and time compared to `encode_bytes`.
+    /// Note that the encoded result is not memory-comparable.
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()>;
+}
+
+impl<T: NumberEncoder> CompactByteEncoder for T {
+    #[inline]
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()> {
+        self.write_var_i64(data.len() as i64)?;
+        self.write_all_bytes(data)
+    }
+}
+
+impl CompactByteEncoder for std::fs::File {
+    #[inline]
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()> {
+        use std::io::Write;
+        let mut buf = vec![0; super::number::MAX_VARINT64_LENGTH];
+        let written = NumberCodec::encode_var_i64(&mut buf, data.len() as i64);
+        self.write_all(&buf[..written])?;
+        self.write_all(data)?;
+        Ok(())
+    }
+}
+
+pub trait CompactByteDecoder {
+    /// Decodes bytes which are encoded by `write_compact_bytes` before.
+    fn read_compact_bytes(&mut self) -> Result<Vec<u8>>;
+
+    /// Gets the first encoded bytes' length in compactly encoded data.
+    ///
+    /// Compact-encoding includes a VarInt encoded length prefix (1 ~ 9 bytes) and N bytes payload.
+    /// This function gets the total bytes length of compact-encoded data, including the length prefix.
+    ///
+    /// Note:
+    ///     - This function won't check whether the bytes are encoded correctly.
+    ///     - There can be multiple compact-encoded data, placed one by one. This function only returns
+    ///       the length of the first one.
+    fn encoded_compact_len(&mut self) -> usize;
+}
+
+impl<T: NumberDecoder> CompactByteDecoder for T {
+    fn read_compact_bytes(&mut self) -> Result<Vec<u8>> {
```

Comment:
for this one, may be we can return reference to avoid one copy.

Reply:
the caller of this function need a Vec, [Datum::Bytes](https://github.com/tikv/tikv/blob/cb0e3046ca315ce00964b9bf11bd210c08f4ccdf/src/coprocessor/codec/datum.rs#L833) and [mvcc::Lock](https://github.com/tikv/tikv/blob/56a9a37a7c3357f9ced12ce6534f128b750c43dc/src/storage/mvcc/lock.rs#L106),  if we change result type to a reference, then we need to casting after invoke this function

Reply:
Let's cast it :) As an underlying function, it should avoid opt-in these cost.

Reply:
currently my implemention `impl<T: Read> CompactByteDecoder for std::io::BufReader<T>`
can't return reference, the result is a local variable..
https://github.com/niedhui/tikv/blob/52186751fff5ccfd7fc2573d82928c080c9c0733/components/codec/src/byte.rs#L493